File: data_structures/codewars_c/4 kyu/roman_to_arabic.cpp
#include <iostream>
#include <string>

using namespace std;

int decode(char letter)
{
  switch (letter)
  {
    case 'M':
      return 1000;
    case 'D':
      return 500;
    case 'C':
      return 100;
    case 'L':
      return 50;
    case 'X':
      return 10;
    case 'V':
      return 5;
    case 'I':
      return 1;
  }
}

int solution(string roman) {
  int res = 0;
  for (int i = 0; i < roman.length(); ++i)
    if (i < roman.length() - 1 && decode(roman[i]) < decode(roman[i + 1])) {
      res -= decode(roman[i]);
    } else res += decode(roman[i]);

  return res;
}

int main() {
  cout << solution("XXI") << endl;
  return 0;
}

File: data_structures/codewars_c/4 kyu/roman_to_arabic.js
const dict = {I: 1, V: 5, X: 10, L: 50, C: 100, D: 500, M: 1000};

const solution = (roman) => {
  let res = 0;

  const objectValue = (el) => {
    for (let prop in dict) {
        if (prop == el) return dict[prop];
      }
  }

  roman = roman.split('');
  for (let i = 0; i < roman.length; i++) {
    if (objectValue(roman[i]) < objectValue(roman[i+1])) {
      res += objectValue(roman[i+1]) - objectValue(roman[i]);
      i++;
    } else res += objectValue(roman[i]);
  }
  return res;
}

console.log(solution('XXI'));
